### DISCUSSION TITLE
Serious Memory Leak: Nested x-ifs and x-fors do not clean up Watchers and Effects, causing memory to grow indefinitely

### DISCUSSION THREAD

**Discussion by mgschoen** [alpinejs/alpine#4299] [7. Bugs (archived)]
Following https://github.com/alpinejs/alpine/discussions/4287 I did some additional research and discovered the following, pretty serious memory leak related to the `x-if` and `x-for` directives.

Consider a situation in which two (or more) x-ifs or x-fors are nested. Whenever the expression of the outer directive changes in such a way that content gets removed, watchers and effects that are defined inside the inner directive are not cleaned up and stay in memory indefinitely.

Minimal example:

```html
<div x-data="{ show: true }">
  <input type="checkbox" x-model="show" />
  <template x-if="show">
    <div>
        <template x-if="true">
          <div x-effect="console.log('toggle checkbox and I will get multiplied', show)">Content</div>
        </template>
    </div>
  </template>
</div>

```

This behavior can be reproduced in a variety of combinations, using `x-if`, `x-for`, `x-effect` and `$watch`. The `cleanup()` method of custom directives does not get called either. Here are some examples:

- [x-if + x-if + x-effect](https://awesomealpine.com/play/?coreplugins=0&script=LyogKi8K&html=PGRpdiB4LWRhdGE9Insgc2hvdzogdHJ1ZSB9Ij4KICA8aW5wdXQKICAgIHR5cGU9ImNoZWNrYm94IgogICAgeC1tb2RlbD0ic2hvdyIKICAvPgogIDx0ZW1wbGF0ZSB4LWlmPSJzaG93Ij4KICAgIDxkaXY-CiAgICAgIDx0ZW1wbGF0ZSB4LWlmPSJ0cnVlIj4KICAgICAgICA8ZGl2CiAgICAgICAgICB4LWVmZmVjdD0iY29uc29sZS5sb2coJ3RvZ2dsZSBjaGVja2JveCBhbmQgSSB3aWxsIGdldCBtdWx0aXBsaWVkJywgc2hvdykiCiAgICAgICAgPgogICAgICAgICAgQ29udGVudAogICAgICAgIDwvZGl2PgogICAgICA8L3RlbXBsYXRlPgogICAgPC9kaXY-CiAgPC90ZW1wbGF0ZT4KPC9kaXY-Cg&tw=1&panel=script) (same as above)
- [x-if + x-for + x-effect](https://awesomealpine.com/play/?coreplugins=0&script=LyogKi8K&html=PGRpdiB4LWRhdGE9Insgc2hvdzogdHJ1ZSB9Ij4KICA8aW5wdXQKICAgIHR5cGU9ImNoZWNrYm94IgogICAgeC1tb2RlbD0ic2hvdyIKICAvPgogIDx0ZW1wbGF0ZSB4LWlmPSJzaG93Ij4KICAgIDxkaXY-CiAgICAgIDx0ZW1wbGF0ZSB4LWZvcj0iaXRlbSBpbiBbMV0iPgogICAgICAgIDxkaXYKICAgICAgICAgIHgtZWZmZWN0PSJjb25zb2xlLmxvZygndG9nZ2xlIGNoZWNrYm94IGFuZCBJIHdpbGwgZ2V0IG11bHRpcGxpZWQnLCBzaG93KSIKICAgICAgICA-CiAgICAgICAgICBDb250ZW50CiAgICAgICAgPC9kaXY-CiAgICAgIDwvdGVtcGxhdGU-CiAgICA8L2Rpdj4KICA8L3RlbXBsYXRlPgo8L2Rpdj4K&tw=1)
- [x-if + x-if + $watch](https://awesomealpine.com/play/?coreplugins=0&script=LyogKi8K&html=PGRpdiB4LWRhdGE9Insgc2hvdzogdHJ1ZSB9Ij4KICA8aW5wdXQKICAgIHR5cGU9ImNoZWNrYm94IgogICAgeC1tb2RlbD0ic2hvdyIKICAvPgogIDx0ZW1wbGF0ZSB4LWlmPSJzaG93Ij4KICAgIDxkaXY-CiAgICAgIDx0ZW1wbGF0ZSB4LWlmPSJ0cnVlIj4KICAgICAgICA8ZGl2CiAgICAgICAgICB4LWVmZmVjdD0iJHdhdGNoKCdzaG93JywgKCkgPT4gY29uc29sZS5sb2coJ3RvZ2dsZSBjaGVja2JveCBhbmQgSSB3aWxsIGdldCBtdWx0aXBsaWVkJykpIgogICAgICAgID4KICAgICAgICAgIENvbnRlbnQKICAgICAgICA8L2Rpdj4KICAgICAgPC90ZW1wbGF0ZT4KICAgIDwvZGl2PgogIDwvdGVtcGxhdGU-CjwvZGl2Pgo&tw=1)
- [x-if + x-for + $watch](https://awesomealpine.com/play/?coreplugins=0&script=LyogKi8K&html=PGRpdiB4LWRhdGE9Insgc2hvdzogdHJ1ZSB9Ij4KICA8aW5wdXQKICAgIHR5cGU9ImNoZWNrYm94IgogICAgeC1tb2RlbD0ic2hvdyIKICAvPgogIDx0ZW1wbGF0ZSB4LWlmPSJzaG93Ij4KICAgIDxkaXY-CiAgICAgIDx0ZW1wbGF0ZSB4LWZvcj0iaXRlbSBpbiBbMV0iPgogICAgICAgIDxkaXYKICAgICAgICAgIHgtZWZmZWN0PSIkd2F0Y2goJ3Nob3cnLCAoKSA9PiBjb25zb2xlLmxvZygndG9nZ2xlIGNoZWNrYm94IGFuZCBJIHdpbGwgZ2V0IG11bHRpcGxpZWQnKSkiCiAgICAgICAgPgogICAgICAgICAgQ29udGVudAogICAgICAgIDwvZGl2PgogICAgICA8L3RlbXBsYXRlPgogICAgPC9kaXY-CiAgPC90ZW1wbGF0ZT4KPC9kaXY-Cg&tw=1&panel=script)
- [x-for + x-if + x-effect](https://awesomealpine.com/play/?coreplugins=0&script=LyogKi8K&html=PGRpdiB4LWRhdGE9InsgaXRlbXM6IFsxXSB9Ij4KICA8YnV0dG9uIEBjbGljaz0iaXRlbXMubGVuZ3RoID8gaXRlbXMgPSBbXSA6IGl0ZW1zID0gWzFdIj4KICAgIFRvZ2dsZSBpdGVtcwogIDwvYnV0dG9uPgogIDx0ZW1wbGF0ZSB4LWZvcj0iaXRlbSBpbiBpdGVtcyI-CiAgICA8ZGl2PgogICAgICA8dGVtcGxhdGUgeC1pZj0idHJ1ZSI-CiAgICAgICAgPGRpdgogICAgICAgICAgeC1lZmZlY3Q9ImNvbnNvbGUubG9nKCdjbGljayBidXR0b24gYW5kIEkgd2lsbCBnZXQgbXVsdGlwbGllZCcsIGl0ZW1zKSIKICAgICAgICA-CiAgICAgICAgICBDb250ZW50CiAgICAgICAgPC9kaXY-CiAgICAgIDwvdGVtcGxhdGU-CiAgICA8L2Rpdj4KICA8L3RlbXBsYXRlPgo8L2Rpdj4K&tw=1)
- [x-for + x-if + $watch](https://awesomealpine.com/play/?coreplugins=0&script=LyogKi8K&html=PGRpdiB4LWRhdGE9InsgaXRlbXM6IFsxXSB9Ij4KICA8YnV0dG9uIEBjbGljaz0iaXRlbXMubGVuZ3RoID8gaXRlbXMgPSBbXSA6IGl0ZW1zID0gWzFdIj4KICAgIFRvZ2dsZSBpdGVtcwogIDwvYnV0dG9uPgogIDx0ZW1wbGF0ZSB4LWZvcj0iaXRlbSBpbiBpdGVtcyI-CiAgICA8ZGl2PgogICAgICA8dGVtcGxhdGUgeC1pZj0idHJ1ZSI-CiAgICAgICAgPGRpdgogICAgICAgICAgeC1pbml0PSIkd2F0Y2goJ2l0ZW1zJywgKCkgPT4gY29uc29sZS5sb2coJ2NsaWNrIGJ1dHRvbiBhbmQgSSB3aWxsIGdldCBtdWx0aXBsaWVkJykpIgogICAgICAgID4KICAgICAgICAgIENvbnRlbnQKICAgICAgICA8L2Rpdj4KICAgICAgPC90ZW1wbGF0ZT4KICAgIDwvZGl2PgogIDwvdGVtcGxhdGU-CjwvZGl2Pgo&tw=1)
- [x-for + x-for + $watch](https://awesomealpine.com/play/?coreplugins=0&script=LyogKi8K&html=PGRpdiB4LWRhdGE9InsgaXRlbXM6IFsxXSB9Ij4KICA8YnV0dG9uIEBjbGljaz0iaXRlbXMubGVuZ3RoID8gaXRlbXMgPSBbXSA6IGl0ZW1zID0gWzFdIj4KICAgIFRvZ2dsZSBpdGVtcwogIDwvYnV0dG9uPgogIDx0ZW1wbGF0ZSB4LWZvcj0iaXRlbSBpbiBpdGVtcyI-CiAgICA8ZGl2PgogICAgICA8dGVtcGxhdGUgeC1mb3I9ImlubmVySXRlbSBpbiBbMV0iPgogICAgICAgIDxkaXYKICAgICAgICAgIHgtaW5pdD0iJHdhdGNoKCdpdGVtcycsICgpID0-IGNvbnNvbGUubG9nKCdjbGljayBidXR0b24gYW5kIEkgd2lsbCBnZXQgbXVsdGlwbGllZCcpKSIKICAgICAgICA-CiAgICAgICAgICBDb250ZW50CiAgICAgICAgPC9kaXY-CiAgICAgIDwvdGVtcGxhdGU-CiAgICA8L2Rpdj4KICA8L3RlbXBsYXRlPgo8L2Rpdj4K&tw=1)
- [x-if + x-if + Custom Directive cleanup() not called](https://awesomealpine.com/play/?coreplugins=0&script=QWxwaW5lLmRpcmVjdGl2ZSgKICAnZGVzdHJveScsCiAgKGVsLCB7IGV4cHJlc3Npb24gfSwgeyBldmFsdWF0ZUxhdGVyLCBjbGVhbnVwIH0pID0-IHsKICAgIGNvbnN0IGV2YWx1YXRlID0gZXZhbHVhdGVMYXRlcihleHByZXNzaW9uKTsKICAgIGNsZWFudXAoKCkgPT4gZXZhbHVhdGUoKSk7CiAgfSwKKTsK&html=PGRpdiB4LWRhdGE9Insgc2hvdzogdHJ1ZSB9Ij4KICA8aW5wdXQKICAgIHR5cGU9ImNoZWNrYm94IgogICAgeC1tb2RlbD0ic2hvdyIKICAvPgogIDx0ZW1wbGF0ZSB4LWlmPSJzaG93Ij4KICAgIDxkaXYgeC1kZXN0cm95PSJjb25zb2xlLmxvZygnSSB3aWxsIGJlIGNhbGxlZCcpIj4KICAgICAgPHRlbXBsYXRlIHgtaWY9InRydWUiPgogICAgICAgIDxkaXYKICAgICAgICAgIHgtZGVzdHJveT0iY29uc29sZS5sb2coJ0kgd2lsbCBOT1QgYmUgY2FsbGVkJykiCiAgICAgICAgPgogICAgICAgICAgQ29udGVudAogICAgICAgIDwvZGl2PgogICAgICA8L3RlbXBsYXRlPgogICAgPC9kaXY-CiAgPC90ZW1wbGF0ZT4KPC9kaXY-Cg&tw=1)

This might seem like a minor issue at first. But please consider larger apps, especially single page apps. Nesting these structures is not too uncommon. In fact, I inspected memory in a larger such app that my team develops. The app quickly piles up hundreds of MB of memory due to all kinds of data not being garbage collected just by navigating from page to page.

We have not yet managed to get to the bottom of this issue in the Alpine codebase. Any insights or potential fixes from people who are more familiar with the codebase would be very much appreciated!

**Comment by ekwoka** [comment]
I'll look into it :)

EDIT: Actually had a in progress PR that solves it entirely, I just got busy and never opened it.

**Comment by ekwoka** [comment]
#4300

**Reply by mgschoen** [reply]
Thanks for the super quick response! I checked out your PR – it eliminates 99.9% of all Memory Leaks in our SPA 🥳

**Reply by ekwoka** [reply]
Whats the 0.1% remaining?

**Reply by mgschoen** [reply]
Not sure yet, but I cannot confidently attribute them to Alpine at all. So unless there's any other evidence, I'd consider this issue solved by your PR
